src/Symfony/Component/Lock/Store/MysqlStore.php

```diff
@@ -0,0 +1,128 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Lock\Store;
+
+use Symfony\Component\Lock\Exception\LockConflictedException;
+use Symfony\Component\Lock\Key;
+use Symfony\Component\Lock\StoreInterface;
+
+/**
+ * MysqlStore is a StoreInterface implementation using MySQL/MariaDB GET_LOCK function.
+ *
+ * @author Jérôme TAMARELLE <jerome@tamarelle.net>
+ */
+class MysqlStore implements StoreInterface
+{
+    private $dsn;
+    private $username;
+    private $password;
+    private $options;
+    private $waitTimeout;
+
+    /**
+     * List of available options:
+     *  * db_username: The username when lazy-connect [default: '']
+     *  * db_password: The password when lazy-connect [default: '']
+     *  * db_connection_options: An array of driver-specific connection options [default: array()]
+     *  * wait_timeout: Time in seconds to wait for a lock to be released. A negative value means infinite. [default: -1].
+     *
+     * @param string $dsn     The connection DSN string
+     * @param array  $options configuration options
+     */
+    public function __construct($dsn, array $options)
+    {
+        $this->dsn = $dsn;
+        $this->username = $options['db_username'] ?? '';
+        $this->password = $options['db_password'] ?? '';
+        $this->options = $options['db_connection_options'] ?? array();
+        $this->waitTimeout = $options['wait_timeout'] ?? -1;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function save(Key $key)
+    {
+        $this->lock($key, false);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function waitAndSave(Key $key)
+    {
+        $this->lock($key, true);
+    }
+
+    private function lock(Key $key, bool $blocking)
+    {
+        // The lock is maybe already acquired.
+        if ($key->hasState(__CLASS__)) {
+            return;
+        }
+
+        // no timeout for impatient
+        $timeout = $blocking ? $this->waitTimeout : 0;
+
+        $connection = new \PDO($this->dsn, $this->username, $this->password, $this->options);
```

Comment:
Here you open a new connection for each lock (is it necessary?). It could trigger a max connections limit. Moreover you don't close theme.

Reply:
The connection is closed automatically when the `PDO` instance is destructed. That's done by removing the `PDOStatement` from the key state.

I chose to open a new connection for each lock to be sure that we don't get the same lock multiple times on the same connection.

> In MySQL 5.7.5, [GET_LOCK()](https://dev.mysql.com/doc/refman/5.7/en/miscellaneous-functions.html#function_get-lock) was reimplemented using the metadata locking (MDL) subsystem and its capabilities were extended. Multiple simultaneous locks can be acquired and GET_LOCK() does not release any existing locks. **It is even possible for a given session to acquire multiple locks for the same name.** Other sessions cannot acquire a lock with that name until the acquiring session releases all its locks for the name. 

But if we allow the injection of a connection, we need to keep this state in-memory.

Reply:
Indeed by sharing PDO you have to to avoid concurrency inside the same process. This side effect should be warned in the documentation IMO

Reply:
I've found this workaround to avoid multiple connections:
```sql
SELECT IF(IS_USED_LOCK(:key) IS NULL, GET_LOCK(:key, :timeout), 0)
```

Commit is coming.

Reply:
> be sure that we don't get the same lock multiple times on the same connection.

Is this really what we want? Looks opinionated to me. flock() doesn't blocks when acquiring twice a lock on a file, I would expect the same here, so that I have more choices in the end (my responsibility to create several connections if I need intra process locking)

Reply:
@nicolas-grekas there is a test for that (https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Lock/Tests/Store/AbstractStoreTest.php#L80). Tools like amphp (ReactPHP) needs locks inside the same process.

Reply:
But same key (remember, a key contains the resource locked + the lock state) can be reused and locked several times inside the same process.

Reply:
Ok indeed